File: 0211-design-add-and-search-words-data-structure/solution.py
class TrieNode:

    def __init__(self):
        self.children = {}
        self.isWordDone = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        cur = self.root
        for x in word:
            if x not in cur.children:
                cur.children[x] = TrieNode()
            cur = cur.children[x]
        cur.isWordDone = True

    def search(self, word: str) -> bool:
        def dfs(index, root):
            cur = root
            for i in range(index, len(word)):

                char = word[i]
                if char == '.':
                    for y in cur.children.values():
                        if dfs(i + 1, y):
                            return True
                    return False
                else: 
                    if char not in cur.children:
                        return False
                    cur = cur.children[char]
            return cur.isWordDone
        return dfs(0, self.root)

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: 0211-design-add-and-search-words-data-structure/solution.cpp
class TrieNode{
    public:
    vector<TrieNode*> nei;
    bool isWord;
    TrieNode(){
        nei = vector<TrieNode*>(26, nullptr);
        isWord = false;
    }
};
class WordDictionary {
public:
    TrieNode* root;
    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode* node = root;
        for(const char& cur : word){
            int index = cur - 'a';
            if(node->nei[index] == nullptr){
                node->nei[index] = new TrieNode();
            }
            node = node->nei[index];
        }
        node->isWord = true;
    }
    
    bool search(string word) {
        TrieNode* node = root;
        return dfs(node, 0, word);
    }
    bool dfs(TrieNode* node, int index, string& word){
        if(index >= word.size()){
            return node->isWord;
        }
        bool res = false;
        if(word[index] == '.'){
            for(int i = 0; i < 26; ++i){
                if(node->nei[i] != nullptr and dfs(node->nei[i], index + 1, word)){
                    return true;
                }
            }
        }else{
            if(node->nei[word[index] - 'a'] != nullptr){
                return dfs(node->nei[word[index] - 'a'], index + 1, word);
            }
        }
        return false;

    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */
